apps/openmw/mwworld/cellstore.cpp

```diff
@@ -1024,4 +1061,83 @@ namespace MWWorld
+        }
+        for (TRechargingItems::iterator it = mRechargingItems.begin(); it != mRechargingItems.end(); ++it)
+        {
+            Ptr ptr = getCurrentPtr(it->first);
+
+            if (ptr.getCellRef().getEnchantmentCharge() == -1
+                    || ptr.getCellRef().getEnchantmentCharge() == it->second)
+                continue;
+
+            static float fMagicItemRechargePerSecond = MWBase::Environment::get().getWorld()->getStore().get<ESM::GameSetting>().find(
+                        "fMagicItemRechargePerSecond")->mValue.getFloat();
+
+            if (ptr.getCellRef().getEnchantmentCharge() <= it->second)
+                ptr.getCellRef().setEnchantmentCharge(std::min (ptr.getCellRef().getEnchantmentCharge() + fMagicItemRechargePerSecond * duration, it->second));
+        }
+    }
+
+    void MWWorld::CellStore::updateRechargingItems()
+    {
+        mRechargingItems.clear();
+
+        for (CellRefList<ESM::Weapon>::List::iterator it (mWeapons.mList.begin()); it!=mWeapons.mList.end(); ++it)
+        {
+            Ptr ptr = getCurrentPtr(&*it);
+            if (!ptr.isEmpty() && ptr.getRefData().getCount() > 0)
+            {
+                checkItem(ptr);
+            }
+        }
+        for (CellRefList<ESM::Armor>::List::iterator it (mArmors.mList.begin()); it!=mArmors.mList.end(); ++it)
+        {
+            Ptr ptr = getCurrentPtr(&*it);
+            if (!ptr.isEmpty() && ptr.getRefData().getCount() > 0)
+            {
+                checkItem(ptr);
+            }
+        }
+        for (CellRefList<ESM::Clothing>::List::iterator it (mClothes.mList.begin()); it!=mClothes.mList.end(); ++it)
+        {
+            Ptr ptr = getCurrentPtr(&*it);
+            if (!ptr.isEmpty() && ptr.getRefData().getCount() > 0)
+            {
+                checkItem(ptr);
+            }
+        }
+        for (CellRefList<ESM::Book>::List::iterator it (mBooks.mList.begin()); it!=mBooks.mList.end(); ++it)
+        {
+            Ptr ptr = getCurrentPtr(&*it);
+            if (!ptr.isEmpty() && ptr.getRefData().getCount() > 0)
+            {
+                checkItem(ptr);
+            }
+        }
+    }
+
+    void MWWorld::CellStore::checkItem(Ptr ptr)
+    {
+        if (ptr.getClass().getEnchantment(ptr).empty())
+            return;
+
+        std::string enchantmentId = ptr.getClass().getEnchantment(ptr);
+        const ESM::Enchantment* enchantment = MWBase::Environment::get().getWorld()->getStore().get<ESM::Enchantment>().search(enchantmentId);
+        if (!enchantment)
+        {
+            Log(Debug::Warning) << "Warning: Can't find enchantment '" << enchantmentId << "' on item " << ptr.getCellRef().getRefId();
+            return;
+        }
+
+        if (enchantment->mData.mType == ESM::Enchantment::WhenUsed
+                || enchantment->mData.mType == ESM::Enchantment::WhenStrikes)
+            mRechargingItems.push_back(std::make_pair(ptr.getBase(), static_cast<float>(enchantment->mData.mCharge)));
```

Comment:
Same emplace_back suggestion here.

It'd be nice to get rid of code duplication in general.